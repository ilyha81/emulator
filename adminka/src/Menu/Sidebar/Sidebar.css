.Sidebar {
    grid-area: sidebar;
    width: 100%;
    height: 100%;
    background: #61dafb;
}