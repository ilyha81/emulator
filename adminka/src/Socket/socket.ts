import io from "socket.io-client";
import {BehaviorSubject} from "rxjs";

/**
 * Типы сообщений.
 */
export enum ESocketMsg {
    MESSAGE = "MESSAGE",
    SHOW_DATA = "SHOW_DATA",
    TEST = "TEST",
}

export type TSocketMessageCB = (data: string) => void;

/**
 * Функция отрисовки данных в соответствующей секции
 * @param data Данные для отрисовки
 */
const drawData = (data: string) => {
    const section = document.getElementById('show-data');
    console.log(data);
    section && (section.innerText = JSON.stringify(data));
};

/**
 * Сам коннект Сокета.
 */
const socket = io.connect('http://localhost:8000', {transports: ['websocket', 'polling']});

/**
 * Создаем подписчикjd.
 */
const MessageSubscriber = new BehaviorSubject<string>("");
const OperationsSubscriber = new BehaviorSubject<string>("");

/**
 * Создаем рассылку.
 */
socket.on(ESocketMsg.MESSAGE, (data: string) => {
    console.warn("SCOKET: incomming message = " + ESocketMsg.MESSAGE);
    MessageSubscriber.next(data)
});
socket.on(ESocketMsg.SHOW_DATA, (data: string) => {
    console.warn("SCOKET: incomming message = " + ESocketMsg.SHOW_DATA);
    OperationsSubscriber.next(data)
});

export const SocketAPI = {
    /**
     * Создание тестовой операции.
     */
    emitTestOperation: () => {
        socket.emit(ESocketMsg.TEST, {
            documents: [{id: 'asd1', docType: 'RPP'}, {id: 'asd2', docType: 'RPP'}],
            cryptoprofileId: "TEST_PROFILE"
        })
    },

    /**
     * Подписка на Получение сообщения.
     * @param {string} cb Колбэк
     */
    subscribeOnHello: (cb: TSocketMessageCB) => MessageSubscriber.subscribe(cb),

    /**
     * Подписка на Получение операций.
     * @param {string} cb Колбэк.
     */
    subscribeOnOperations: (cb: TSocketMessageCB) => OperationsSubscriber.subscribe(cb),

};

