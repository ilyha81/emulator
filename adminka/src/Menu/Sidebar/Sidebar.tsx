import * as Reacr from 'react';
import "./Sidebar.css";

/**
 * Сайдбар с управлением.
 * @constructor
 */
export const Sidebar: Reacr.FC = (props) => {

    return <div className="Sidebar">
        Сайдбар
        {props.children}
    </div>
};