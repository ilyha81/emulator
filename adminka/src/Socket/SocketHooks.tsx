import { useState, useEffect } from 'react';
import {ESocketConnection, SocketAPI} from './socket';

/**
 * Hook subscribes on Socket connection status.
 */
export const useSocketConnectionStatus = ():ESocketConnection  => {
    const [socketStatus, setSocketStatus] = useState<ESocketConnection>(ESocketConnection.INITIAL);

    useEffect(()=>{
        const connected = SocketAPI.subscribeOnConnectionStatus(setSocketStatus);

        return connected.unsubscribe
    },[]);

    return socketStatus
}