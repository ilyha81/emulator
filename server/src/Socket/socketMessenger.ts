import {Timer} from '../Timer/interval';
import * as socketio  from "socket.io";
import * as core from "express-serve-static-core";
import {IOperationView} from "../Operation/operation";
import {Server as Socket} from 'socket.io';

/**
 * Типы сообщений.
 */
export enum ESocketMsg {
    MESSAGE = "MESSAGE",
    SHOW_DATA = "SHOW_DATA",
    TEST = "TEST",
}

/**
 * Проверка занятых портов.
 */
const ports: Set<number> = new Set;

export class SocketMessenger {
    socket: socketio.Server;

    static isPortUsed = (port: number) => ports.has(port);

    constructor (socket: Socket, getOperations: () => IOperationView[]) {
        this.socket = socket;

        this.socket.on('connection', (client: socketio.Socket) => {
            console.log('Клиент подключился');

            client.emit(ESocketMsg.MESSAGE, "ПОДКЛЮЧЕНИЕ УСПЕШНО !");
        });
    }

    /**
     * Обработка приема сообщений.
     * @param type
     * @param cb
     */
    on = (type: string, cb: (client: any) => any) => {
        this.socket.on(type, cb)
    };

    emitMsg = (type: string, msg: object) => {
        if (typeof msg === "string") {
            this.socket.emit(type, msg);
        } else {
            this.socket.emit(type, JSON.stringify(msg))
        }
    }
}
