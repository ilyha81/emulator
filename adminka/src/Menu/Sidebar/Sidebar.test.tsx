import * as React from 'react';
import {cleanup, render, screen, fireEvent} from '@testing-library/react';

import {Sidebar} from './Sidebar';
import {ESocketConnection} from "../../Socket/socket";

beforeEach(()=> {
    cleanup();
});

describe('Testing Sidebar', () => {

    test('Testing Sidebar render', ()=> {
        const {container} = render(<Sidebar socketStatus={ESocketConnection.INITIAL}/>);

        expect(container).toMatchSnapshot();
    });

    test('Sidebar Name props', () => {
        const name='NAME';
        render(<Sidebar name={name} socketStatus={ESocketConnection.INITIAL}/>);

        expect(screen.getByText(name)).toHaveClass('badge', 'badge-dark');
    });

    test('Sidebar renders children', () => {
        const VASYA = 'VASYA';
        const PETYA = 'PETYA';
        const Child:React.FunctionComponent<{str: string}> = ({str}) => {
            return <div>{str}</div>
        };

        render(<Sidebar socketStatus={ESocketConnection.INITIAL}>
            <Child str={VASYA}/>
            <Child str={PETYA}/>
        </Sidebar>);

        screen.getByText(VASYA);
        screen.getByText(PETYA);
    });

    test('Sidebar has Test Operation button', () => {
        render(<Sidebar socketStatus={ESocketConnection.INITIAL}/>);

        expect(screen.getByText('Create Test Operation')).toHaveClass('btn', 'btn-lg', 'btn-block', 'btn-primary');
    });

    test('Sidebar Test Operation button disabled-enabled', () => {
        const { rerender } = render(<Sidebar socketStatus={ESocketConnection.INITIAL}/>);
        expect(screen.getByText('Create Test Operation')).toBeDisabled();

        rerender(<Sidebar socketStatus={ESocketConnection.CONNECTED}/>);
        expect(screen.getByText('Create Test Operation')).not.toBeDisabled();

        rerender(<Sidebar socketStatus={ESocketConnection.DISCONNECTED}/>);
        expect(screen.getByText('Create Test Operation')).toBeDisabled();

        rerender(<Sidebar socketStatus={ESocketConnection.FAILED}/>);
        expect(screen.getByText('Create Test Operation')).toBeDisabled();
    });

    test('Sidebar Test Operation button onClick', async () => {
        const mockCb = jest.fn();
        render(<Sidebar socketStatus={ESocketConnection.CONNECTED}/>);
        const operationButton = screen.getByText('Create Test Operation');
        expect(operationButton).not.toBeDisabled();

        operationButton.onclick = mockCb;
        fireEvent.click(operationButton);
        expect(mockCb).toHaveBeenCalledTimes(1);
        fireEvent.click(operationButton);
        expect(mockCb).toHaveBeenCalledTimes(2);
    })
});