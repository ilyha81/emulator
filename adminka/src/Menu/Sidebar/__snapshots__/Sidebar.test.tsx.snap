// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Sidebar Testing Sidebar render 1`] = `
<div>
  <div
    class="Sidebar container pt-2"
  >
    <button
      class="btn btn-block btn-lg btn-primary"
      disabled=""
    >
      Create Test Operation
    </button>
  </div>
</div>
`;
